File: python/src/dionysus/api/oauth.py
"""Machine OAuth-style token API routes."""

from datetime import UTC, datetime
from json import JSONDecodeError
from typing import Any

from fastapi import APIRouter, HTTPException, Request, Response, status
from pydantic import BaseModel, ConfigDict, ValidationError

from dionysus.identity.machines import exchange_machine_client_secret

router = APIRouter(prefix="/api/oauth", tags=["oauth"])

CLIENT_CREDENTIALS_GRANT = "client_credentials"
BEARER_AUTH_SCHEME = "bearer"


class TokenRequest(BaseModel):
    """Machine token exchange request body."""

    model_config = ConfigDict(extra="forbid")

    grant_type: str
    client_id: str
    client_secret: str


class TokenResponse(BaseModel):
    """Machine bearer and refresh token response body."""

    model_config = ConfigDict(extra="forbid")

    access_token: str
    token_type: str
    expires_in: int
    refresh_token: str
    refresh_expires_in: int


def _invalid_client_credentials() -> HTTPException:
    """Return the generic machine credential exchange failure.

    Returns:
        A 401 HTTP exception that does not reveal whether the client ID or
        client secret failed verification.
    """

    return HTTPException(
        status_code=status.HTTP_401_UNAUTHORIZED,
        detail="Invalid client credentials",
        headers={"WWW-Authenticate": "Bearer"},
    )


def _invalid_token_request() -> HTTPException:
    return HTTPException(
        status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
        detail="Invalid token request",
    )


async def _token_request_from_request(request: Request) -> TokenRequest:
    content_type = request.headers.get("content-type", "").partition(";")[0].strip().lower()
    try:
        payload: Any
        if content_type in {
            "application/x-www-form-urlencoded",
            "multipart/form-data",
        }:
            payload = dict(await request.form())
        else:
            payload = await request.json()
        return TokenRequest.model_validate(payload)
    except (JSONDecodeError, ValidationError):
        raise _invalid_token_request() from None


@router.post("/token", response_model=TokenResponse)
async def create_machine_token(request: Request, response: Response) -> TokenResponse:
    """Exchange machine client credentials for bearer tokens.

    Args:
        request: Incoming request containing application state.
        response: Outgoing response whose headers are updated for successful
            token exchanges.

    Returns:
        An OAuth2-ish bearer access token response with a rotating refresh
        token.

    Raises:
        HTTPException: If the grant type is unsupported or credentials fail
            generic verification.
    """

    credentials = await _token_request_from_request(request)
    if credentials.grant_type != CLIENT_CREDENTIALS_GRANT:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Unsupported grant_type",
        )

    session_factory = request.app.state.session_factory
    settings = request.app.state.settings
    access_expires_minutes = settings.machine_access_token_expires_minutes
    refresh_expires_minutes = settings.machine_refresh_token_expires_minutes

    with session_factory() as db_session:
        token_pair = exchange_machine_client_secret(
            db_session,
            client_id=credentials.client_id,
            client_secret=credentials.client_secret,
            now=datetime.now(UTC),
            access_expires_in_minutes=access_expires_minutes,
            refresh_expires_in_minutes=refresh_expires_minutes,
        )
        if token_pair is None:
            raise _invalid_client_credentials()
        db_session.commit()

    response.headers["Cache-Control"] = "no-store"
    response.headers["Pragma"] = "no-cache"
    return TokenResponse(
        access_token=token_pair.access_token,
        token_type=BEARER_AUTH_SCHEME,
        expires_in=access_expires_minutes * 60,
        refresh_token=token_pair.refresh_token,
        refresh_expires_in=refresh_expires_minutes * 60,
    )

File: backend/internal/http/oauth.go
package httpapi

import (
	"encoding/json"
	"net/http"
	"strings"
	"time"

	"github.com/go-chi/chi/v5"
	"github.com/invacuation/dionysus/backend/internal/config"
	"github.com/invacuation/dionysus/backend/internal/identity"
)

const (
	clientCredentialsGrant = "client_credentials"
	bearerAuthScheme       = "bearer"
)

type tokenRequest struct {
	GrantType    string `json:"grant_type"`
	ClientID     string `json:"client_id"`
	ClientSecret string `json:"client_secret"`
}

type tokenResponse struct {
	AccessToken      string `json:"access_token"`
	TokenType        string `json:"token_type"`
	ExpiresIn        int    `json:"expires_in"`
	RefreshToken     string `json:"refresh_token"`
	RefreshExpiresIn int    `json:"refresh_expires_in"`
}

func mountOAuthRoutes(router chi.Router, settings config.Settings, deps Dependencies) {
	router.Post("/api/oauth/token", func(w http.ResponseWriter, r *http.Request) {
		createMachineToken(w, r, settings, deps)
	})
}

// Create a new machine token, given a valid client ID and secret.
// This is used for machine-to-machine authentication.
func createMachineToken(w http.ResponseWriter, r *http.Request, settings config.Settings, deps Dependencies) {
	if deps.DB == nil {
		writeError(w, http.StatusServiceUnavailable, "Database unavailable")
		return
	}
	credentials, ok := tokenRequestFromRequest(w, r)
	if !ok {
		return
	}
	if credentials.GrantType != clientCredentialsGrant {
		writeError(w, http.StatusBadRequest, "Unsupported grant_type")
		return
	}

	tokenPair, err := identity.ExchangeMachineClientSecret(r.Context(), deps.DB, identity.MachineClientSecretExchange{
		ClientID:                credentials.ClientID,
		ClientSecret:            credentials.ClientSecret,
		Now:                     time.Now().UTC(),
		AccessExpiresInMinutes:  settings.MachineAccessTokenExpiresMinutes,
		RefreshExpiresInMinutes: settings.MachineRefreshTokenExpiresMinutes,
	})
	if err != nil {
		writeError(w, http.StatusInternalServerError, "Unable to create machine token!")
		return
	}
	if tokenPair == nil {
		w.Header().Set("WWW-Authenticate", "Bearer")
		writeError(w, http.StatusUnauthorized, "Invalid client credentials")
		return
	}

	w.Header().Set("Cache-Control", "no-store")
	w.Header().Set("Pragma", "no-cache")
	writeJSON(w, http.StatusOK, tokenResponse{
		AccessToken:      tokenPair.AccessToken,
		TokenType:        bearerAuthScheme,
		ExpiresIn:        settings.MachineAccessTokenExpiresMinutes * 60,
		RefreshToken:     tokenPair.RefreshToken,
		RefreshExpiresIn: settings.MachineRefreshTokenExpiresMinutes * 60,
	})
}

// This method retrieves token request (i.e. client ID and secret) from either a JSON body or form values, depending on the Content-Type header of the request.
func tokenRequestFromRequest(w http.ResponseWriter, r *http.Request) (tokenRequest, bool) {
	contentType := strings.ToLower(strings.TrimSpace(strings.Split(r.Header.Get("Content-Type"), ";")[0]))
	if contentType == "application/x-www-form-urlencoded" || contentType == "multipart/form-data" {
		if err := r.ParseForm(); err != nil {
			writeError(w, http.StatusUnprocessableEntity, "Invalid token request")
			return tokenRequest{}, false
		}
		return tokenRequest{
			GrantType:    r.Form.Get("grant_type"),
			ClientID:     r.Form.Get("client_id"),
			ClientSecret: r.Form.Get("client_secret"),
		}, true
	}

	var request tokenRequest
	if err := json.NewDecoder(r.Body).Decode(&request); err != nil {
		writeError(w, http.StatusUnprocessableEntity, "Invalid token request")
		return tokenRequest{}, false
	}
	if request.GrantType == "" || request.ClientID == "" || request.ClientSecret == "" {
		writeError(w, http.StatusUnprocessableEntity, "Invalid token request")
		return tokenRequest{}, false
	}
	return request, true
}

func writeJSON(w http.ResponseWriter, status int, body any) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status)
	_ = json.NewEncoder(w).Encode(body)
}

func writeError(w http.ResponseWriter, status int, detail string) {
	writeJSON(w, status, struct {
		Detail string `json:"detail"`
	}{Detail: detail})
}
